spock-core/src/main/java/org/spockframework/util/JavaProcessThreadDumpCollector.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.spockframework.util;
+
+import org.spockframework.runtime.SpockException;
+import org.spockframework.runtime.extension.builtin.ThreadDumpUtilityType;
+
+import java.io.*;
+import java.lang.management.ManagementFactory;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+
+public interface JavaProcessThreadDumpCollector {
+
+  void appendThreadDumpOfCurrentJvm(StringBuilder builder) throws IOException, InterruptedException;
+
+  static JavaProcessThreadDumpCollector create(ThreadDumpUtilityType utility) {
+    try {
+      return new FunctionalJavaProcessThreadDumpCollector(utility);
+    } catch (Exception e) {
+      ByteArrayOutputStream stream = new ByteArrayOutputStream();
+      e.printStackTrace(new PrintStream(stream));
+      System.out.printf("Thread dump capturing is not available: " + stream);
+      return NO_OP;
+    }
+  }
+
+  JavaProcessThreadDumpCollector NO_OP = __ -> {
+  };
+
+  class FunctionalJavaProcessThreadDumpCollector implements JavaProcessThreadDumpCollector {
+
+    private static final String JAVA_HOME_SYS_PROP = "java.home";
+
+    private final ThreadDumpUtilityType utility;
+    private final List<String> command;
+
+    public FunctionalJavaProcessThreadDumpCollector(ThreadDumpUtilityType utility) {
+      long pid = currentProcessId();
+      Path utilityPath = getUtilityCommandPath(utility);
+
+      this.utility = utility;
+      this.command = utility == ThreadDumpUtilityType.JSTACK
+        ? Arrays.asList(utilityPath.toString(), Long.toString(pid))
+        : Arrays.asList(utilityPath.toString(), Long.toString(pid), "Thread.print");
```

Comment:
I would move this into the `ThreadDumpUtilityType.getCommand(getJavaHome(), pid)`.
Furthermore, I'd also move introduce an interface `ThreadDumpUtility` with that method and have the `ThreadDumpUtilityType` enum implement it. This way one could inject ones own implementation.